source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -379,6 +380,27 @@ void HappyEyeballsConnectionImpl::dumpState(std::ostream& os, int indent_level)
   }
 }
 
+std::vector<Address::InstanceConstSharedPtr>
+HappyEyeballsConnectionImpl::sortAddresses(const std::vector<Address::InstanceConstSharedPtr>& in) {
+  std::vector<Address::InstanceConstSharedPtr> address_list = in;
+  for (auto current = address_list.begin(); current != address_list.end(); ++current) {
+    // Find the first address with a different address family than the current address.
+    auto it = std::find_if(current, address_list.end(), [&](const auto& val) {
+      return (val->type() != Address::Type::Ip || (*current)->type() != Address::Type::Ip ||
+              (*current)->ip()->version() != val->ip()->version());
+    });
+    // If there are no more addresses with different families the sorting is finished.
+    if (it == address_list.end()) {
+      break;
+    }
+    // Bubble the address up to the current position.
+    auto start = std::make_reverse_iterator(it + 1);
+    auto end = std::make_reverse_iterator(current + 1);
+    std::rotate(start, start + 1, end);
+  }
+  return address_list;
```

Comment:
I'd already had concernes about handling "infinite" addresses, but I have even more if we're doing N^2 work.  Can we cap returned addresses at some limit (10?) to limit work done?  Then we can just recreate the array with iterators to v4 and v6 and alternating?

Reply:
Great point about N^2. As discussed offline, switched to an O(N) implementation but did not truncate the address list.